### PR TITLE
Implements linear acquisition model norm computation

### PR DESCRIPTION
**Pull request by evgueni-ovtchinnikov**
Addresses #804.

Uses Jacobi-conjugated gradients algorithm for computing the norm of (the linear part of) an acquisition model.

### CODE REVIEW

File: src/xSTIR/cSTIR/tests/test1.cpp [hunk 1]
```diff
@@ -27,11 +27,13 @@ limitations under the License.
 \author Richard Brown
 \author SyneRBI
 */
+#include <cmath>
```
Comment (KrisThielemans) [inline src/xSTIR/cSTIR/tests/test1.cpp:30]: possibly not needed but fine of course

File: src/xSTIR/cSTIR/tests/test1.cpp [hunk 2]
```diff
@@ -125,6 +127,11 @@ int test1()
 		am.set_background_term(sptr_b);
 		am.set_up(sptr_ad, sptr_id);
 
+		// compute the norm of the linear part of the acquisition model
+		std::cout << "computing the norm of the linear part of the acquisition model...\n";
+		float lmd = am.norm();
+		std::cout << "acquisition model operator norm: " << lmd << '\n';
+
```
Comment (KrisThielemans) [inline src/xSTIR/cSTIR/tests/test1.cpp:134]: can we add a test on the actual norm? Presumably something like
```
y = Ax
norm(y)<=norm(A)*norm(x)
```
for some random x.

File: src/xSTIR/cSTIR/include/sirf/STIR/stir_x.h [hunk 3]
```diff
@@ -33,9 +33,11 @@ limitations under the License.
 \author SyneRBI
 */
 
+#include <cmath>
 #include <stdlib.h>
 
 #include "sirf/STIR/stir_data_containers.h"
+#include "sirf/common/jcg.h"
```
Comment (KrisThielemans) [inline src/xSTIR/cSTIR/include/sirf/STIR/stir_x.h:40]: please rename file to JacobiCG.h

File: src/xSTIR/cSTIR/include/sirf/STIR/stir_x.h [hunk 4]
```diff
@@ -398,7 +437,8 @@ The actual algorithm is described in
 			sptr_asm_.reset();
 			//sptr_normalisation_.reset();
 		}
-		stir::shared_ptr<const PETAcquisitionModel> linear_acq_mod_sptr() const
+//		stir::shared_ptr<const PETAcquisitionModel> linear_acq_mod_sptr() const
```
Comment (KrisThielemans) [inline src/xSTIR/cSTIR/include/sirf/STIR/stir_x.h:440]: why do we need to get rid of the `const`? It seems to open us up for problems.

File: src/common/include/sirf/common/jcg.h [hunk 5]
```diff
@@ -0,0 +1,123 @@
+#pragma once
+
+#include <cmath>
+#include <iostream>
+
+#include <boost/shared_ptr.hpp>
+
+template<class vector>
```
Comment (KrisThielemans) [inline src/common/include/sirf/common/jcg.h:8]: maybe confusing to use `vector` (at first glance anyway). what about calling it `vector_type` or `vectorT` or something

File: src/common/include/sirf/common/jcg.h [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+class anOperator {
```
Comment (KrisThielemans) [inline src/common/include/sirf/common/jcg.h:9]: new base class shouldn't be defined in `jcg.h` but have its own file, and doxygen.

I'm not convinced about the name `anOperator`. I don't think other classes have `a` or `an` (although I'm not sure!). `Operator` seems better to me.

I guess in future have to think what to stick in here, but that can easily be postponed.

File: src/common/include/sirf/common/jcg.h [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+
+		value s = sqrt(x.dot(x));
+		x.scale(s);
+		std::shared_ptr<vector> sptr_Ax = A(x);
+		vector& Ax = *sptr_Ax;
+
+		for (int it = 0; it < nit_; it++) {
+			lmd = Ax.dot(x);
+			y.axpby(1.0, Ax, -lmd, x);
+			std::cout << it << ": " << lmd << '\n';
```
Comment (KrisThielemans) [inline src/common/include/sirf/common/jcg.h:52]: best to remove this later, or at least say more

File: src/common/include/sirf/common/jcg.h [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+			eigh2_(a, mu, u, v);
+			z.axpby(u[0], x, u[1], y);
+			x.axpby(v[0], x, v[1], y);
+			Az.axpby(u[0], Ax, u[1], Ay);
+			Ax.axpby(v[0], Ax, v[1], Ay);
+			lmd = mu[1];
+		}
+		return lmd;
+	}
+	void try_eigh2(const value a[4], value lmd[2], value x[2], value y[2])
```
Comment (KrisThielemans) [inline src/common/include/sirf/common/jcg.h:84]: public? What does it do (in any case document)

**Review by KrisThielemans — CHANGES_REQUESTED** [review]
looks good. house-keeping comments before you merge. I didn't check the math though. Main thing is to add a test.

**Comment by KrisThielemans** [comment]
@epapoutsellis what do you think?

File: src/xSTIR/cSTIR/include/sirf/STIR/stir_x.h [hunk 9]
```diff
@@ -334,6 +336,43 @@ The actual algorithm is described in
 
 	class PETAcquisitionModel {
 	public:
+		class BFOperator : public anOperator<STIRImageData> {
```
[the remark this answers is not in the document]
Comment (evgueni-ovtchinnikov) [reply src/xSTIR/cSTIR/include/sirf/STIR/stir_x.h]: done

File: src/xSTIR/cSTIR/tests/test1.cpp (hunk 2 above)
Comment (evgueni-ovtchinnikov) [reply src/xSTIR/cSTIR/tests/test1.cpp]: done

File: src/xSTIR/cSTIR/include/sirf/STIR/stir_x.h (hunk 3 above)
Comment (evgueni-ovtchinnikov) [reply src/xSTIR/cSTIR/include/sirf/STIR/stir_x.h]: done

File: src/xSTIR/cSTIR/include/sirf/STIR/stir_x.h (hunk 4 above)
Comment (evgueni-ovtchinnikov) [reply src/xSTIR/cSTIR/include/sirf/STIR/stir_x.h]: `forward` and `backward` were not `const` - they are now

File: src/common/include/sirf/common/jcg.h (hunk 5 above)
Comment (evgueni-ovtchinnikov) [reply src/common/include/sirf/common/jcg.h]: done

File: src/common/include/sirf/common/jcg.h (hunk 6 above)
Comment (evgueni-ovtchinnikov) [reply src/common/include/sirf/common/jcg.h]: done

File: src/common/include/sirf/common/jcg.h (hunk 7 above)
Comment (evgueni-ovtchinnikov) [reply src/common/include/sirf/common/jcg.h]: commented out for now, ideally `verbosity` needed

File: src/common/include/sirf/common/jcg.h (hunk 8 above)
Comment (evgueni-ovtchinnikov) [reply src/common/include/sirf/common/jcg.h]: sorry, debugging leftover, removed

File: src/xSTIR/cSTIR/include/sirf/STIR/stir_x.h (hunk 4 above)
Comment (evgueni-ovtchinnikov) [reply src/xSTIR/cSTIR/include/sirf/STIR/stir_x.h]: (I was in a hurry to make this pull request before my annual leave next week, and was not sure imposing `const`-ness would go smoothly - it certainly will not on MR side - so went for quick fix)

File: src/xSTIR/cSTIR/tests/test1.cpp [hunk 10]
```diff
@@ -214,11 +216,31 @@ int test1()
...
+		// compute the norm of the linear part of the acquisition model
+		std::cout << "\ncomputing the norm of the linear part of the acquisition model...\n";
+		float am_norm = am.norm();
+
+		std::cout << "\nchecking the acquisition model norm:\n";
+		std::cout << "acquisition model norm: |A| = " << am_norm << '\n';
+		std::cout << "image data x norm: |x| = " << im_norm << '\n';
+		std::cout << "simulated acquisition data norm: |A(x)| = " << sim_norm << '\n';
+		std::cout << "checking that |A(x)| < |A| |x|: ";
+		bool ok = sim_norm < am_norm*sim_norm;
```
Comment (KrisThielemans) [inline src/xSTIR/cSTIR/tests/test1.cpp:229]: should be `<=` (you could be unlucky!). example: `x=0`

**Review by KrisThielemans — CHANGES_REQUESTED** [review]
great.  thanks. one tiny thing. Aside from that, `JacobiCG` and `Operator` need to be in `namespace sirf`.

File: src/xSTIR/cSTIR/include/sirf/STIR/stir_x.h (hunk 4 above)
Comment (KrisThielemans) [reply src/xSTIR/cSTIR/include/sirf/STIR/stir_x.h]: 😄

**Comment by KrisThielemans** [comment]
@paskino @epapoutsellis This seems alright to me. Would be good if you could check if this returns the same norm as the Power-method from CIL. That'd be a nice check...

**Pull request merged**